invisible_cities/evm/pmaps.pyx

```diff
@@ -0,0 +1,272 @@
+                                    if np.any(np.isnan(self.t))  or
+                                       np.any(np.isnan(self.E))
+                                    else True)
+
+
+    def __str__(self):
+        s = """Peak(samples = {0:d} width = {1:8.1f} mus , energy = {2:8.1f} pes
+        height = {3:8.1f} pes tmin-tmax = {4} mus """.format(self.number_of_samples,
+        self.width / units.mus, self.total_energy, self.height,
+        (self.tmin_tmax * (1 / units.mus)))
+        return s
+
+    def __repr__(self):
+        return self.__str__()
+
+
+cdef class S12:
+    """Base class representing an S1/S2 signal
+    The S12 attribute is a dictionary s12
+    {i: Peak(t,E)}, where i is peak number.
+    The notation _s12 is intended to make this
+    class private (public classes s1 and s2 will
+    extend it).
+    The rationale to use s1 and s2 rather than a single
+    class s12 to represent both s1 and s2 is that, although
+    structurally identical, s1 and s2 represent quite
+    different objects. In Particular an s2si is constructed
+    with a s2 not a s1.
+    An s12 is represented as a dictinary of Peaks.
+
+    """
+    def __init__(self, dict s12d):
+
+        cdef int peak_no
+        cdef np.ndarray[double, ndim=1] t
+        cdef np.ndarray[double, ndim=1] E
+        self.peaks = {}
+
+        #print('s12d ={}'.format(s12d))
+        for peak_no, (t, E) in s12d.items():
+            #print('t ={}'.format(t))
+            #print('E ={}'.format(E))
+            assert len(t) == len(E)
+            #p = Peak(t,E)
+            #print('peak = {}'.format(p))
+
+            self.peaks[peak_no] =  Peak(t, E)
+
+    property number_of_peaks:
+         def __get__(self): return len(self.peaks)
+
+    cpdef peak_collection(self):
+        try:
+            return tuple(self.peaks.keys())
+        except KeyError:
+            raise PeakNotFound
+
+    cpdef peak_waveform(self, int peak_number):
+        try:
+            return self.peaks[peak_number]
+        except KeyError:
+             raise PeakNotFound
+
+    cpdef store(self, table, event_number):
+        row = table.row
+        for peak_number, peak in self.peaks.items():
+            for t, E in zip(peak.t, peak.E):
+                row["event"] = event_number
+                row["peak"]  =  peak_number
+                row["time"]  = t
+                row["ene"]   = E
+                row.append()
+
+
+cdef class S1(S12):
+    def __init__(self, s1d):
+        self.s1d = s1d
+        super(S1,self).__init__(s1d)
+
+    def __str__(self):
+        s =  "S1 (number of peaks = {})\n".format(self.number_of_peaks)
+        s2 = ['peak number = {}: {} \n'.format(i,
+                                    self.peak_waveform(i)) for i in self.peaks]
+        return  s + ''.join(s2)
+
+    def __repr__(self):
+        return self.__str__()
+
+cdef class S2(S12):
+    def __init__(self, s2d):
+        self.s2d = s2d
+        super(S2, self).__init__(s2d)
+
```

Comment:
One minor issue is that if we want a function `manipulate_si` or `manipulate_s12` that interacts either the dictionary in s1 or s2, we have a small problem because the function does not know whether to use si.s1d or si.s2d, right?

Maybe we use s1.d and s2.d to access their dictionaries, which is also more concise. 

In s2si we can leave things as they are since it has two dictionaries. And, since if a function needs to interact with s2si, it needs to know that it is an s2si since s2sid has a different format than s1d and s2d. What do you think @jjgomezcadenas ?